include/onnxruntime/core/common/gpu_profiler_common.h

```diff
@@ -0,0 +1,450 @@
+  // unique_correlation_id - offset == external_correlation_id
+  InlinedHashMap<uint64_t, std::pair<uint64_t, uint64_t>> unique_correlation_id_to_client_offset_;
+
+  // Keyed on tracer_correlation_id -> unique_correlation_id
+  InlinedHashMap<uint64_t, uint64_t> tracer_correlation_to_unique_correlation_;
+
+  // client_id/client_handle -> external_correlation_id -> events
+  InlinedHashMap<uint64_t, std::map<uint64_t, Events>> per_client_events_by_ext_correlation_;
+
+  // Keyed on tracer correlation_id, keeps track of activity records
+  // for which we haven't established the external_correlation_id yet.
+  InlinedHashMap<uint64_t, std::vector<EventRecord>> events_pending_client_mapping_;
+}; /* class GPUTracerManager */
+
+// Base class for a GPU profiler
+template <typename TManager>
+class GPUProfilerBase : public EpProfiler {
+ protected:
+  GPUProfilerBase() = default;
+  virtual ~GPUProfilerBase() {}
+
+  void MergeEvents(std::map<uint64_t, Events>& events_to_merge, Events& events) {
+    Events merged_events;
+
+    auto event_iter = std::make_move_iterator(events.begin());
+    auto event_end = std::make_move_iterator(events.end());
+    for (auto& map_iter : events_to_merge) {
+      auto ts = static_cast<long long>(map_iter.first);
+
+      // find the last occurence of a matching timestamp,
+      // if one exists
+      while (event_iter != event_end &&
+             (event_iter->ts < ts ||
+              (event_iter->ts == ts &&
+               (event_iter + 1) != event_end &&
+               (event_iter + 1)->ts == ts))) {
+        merged_events.emplace_back(*event_iter);
+        ++event_iter;
+      }
+
+      uint64_t increment;
+      uint64_t last_ts;
+      bool copy_op_names = false;
+      std::string op_name;
+      std::string parent_name;
+
+      // Tracers may not use Jan 1 1970 as an epoch for timestamps.
+      // So, we need to adjust the timestamp to something sensible.
+      if (event_iter != event_end && event_iter->ts == ts) {
+        // In this particular case we have located a parent event -- in the main event stream --
+        // for the GPU events. We use the timestamps from that event to adjust timestamps
+        last_ts = event_iter->ts;
+        increment = 1;
+        copy_op_names = true;
+        op_name = event_iter->args["op_name"];
+        parent_name = event_iter->name;
+        merged_events.emplace_back(*event_iter);
+        ++event_iter;
+      } else {
+        // No parent event, let's just set the timestamp based on the
+        // timestamp of the call to EpProfiler->Start()
+        last_ts = ts;
+        increment = 1;
+      }
+
+      for (auto& evt : map_iter.second) {
+        if (copy_op_names) {
+          // If we have found a matching parent event,
+          // then inherit some names from the parent.
+          evt.args["op_name"] = op_name;
+          evt.args["parent_name"] = parent_name;
+        }
+        evt.ts = last_ts + increment;
+        increment += evt.dur;
```

Comment:
Doesn't this should be 'increment += <curr_event_ts> - <prev_event_ts>' somehow? Otherwise, it makes all children events appears to be back to back, that is a critical skew in the trace view and will give the profiler users some illusion about the event timing.

Alternatively, is it viable just substract the timestamp diff between CPU events' time epoch and GPU events' time epoch? That is, we **convert** to unify them directly, instead of compute and try to reconstruct?

Reply:
> Doesn't this should be increment += <curr_event_ts> - <prev_event_ts> somehow? Otherwise, it makes all children events appears to be back to back, that is a critical skew in the trace view and will give the profiler users some illusion about the event timing.

Thanks! This is a much neater way of reconciling the two time streams: we preserve the relative timing between kernel launches.

As for the alternative approach you suggested: that's challenging, primarily because I could not find documentation about the epoch used by roctracer. I'm presuming that it's system boot time, but I could be wrong.